src/patterns/builder/service-example/network-layer/network-layer.impl.ts

```diff
@@ -0,0 +1,43 @@
+import {Builder, EditingHistory, INetworkLayer, Invoice,} from '../contracts'
+import {UserService} from "../services/user-service.impl";
+import {EditingService} from "../services/editing-service.impl";
+import {PaymentService} from "../services/payment-service.impl";
+
+export class NetworkLayer<Added extends INetworkLayer> implements Builder {
+  constructor(protected networkLayer: Added = {} as Added) {
+  }
+
+  static from<S extends INetworkLayer>(services: S): NetworkLayer<S> {
+    return new NetworkLayer(services);
+  }
+
+  build() {
+    return this.networkLayer;
+  }
+
+  createUserService() {
+    return NetworkLayer.from<Added & { userService: UserService }>
+    ({
+      ...this.networkLayer,
+      userService: new UserService(),
+    });
+  }
```

Comment:
I would be great if the type could be inferred form argument, so we do not need to set it manually

```suggestion
  createUserService() {
    return NetworkLayer.from({
      ...this.networkLayer,
      userService: new UserService(),
    });
  }
```